File: frontend/src/services/transaction.ts
import api from "./api";

export type TransactionPayload = {
    title: string;
    amount: number;
    transaction_type: string;
    category: string;
    created_at: string;
};

export async function addTransaction(
    data: TransactionPayload
) {
    const response = await api.post(
        "/api/transactions",
        data
    );

    return response.data;
}

export async function getTransactions() {
    const response = await api.get(
        "/api/transactions"
    );

    return response.data;
}

export async function updateTransaction(
    id: number,
    data: TransactionPayload
) {
    const response = await api.put(
        `/api/transactions/${id}`,
        data
    );

    return response.data;
}

export async function deleteTransaction(id: number) {
    const response = await api.delete(
        `/api/transactions/${id}`
    );

    return response.data;
}

File: backend/app/routers/transaction.py
from fastapi import HTTPException
from typing import List
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.auth.jwt_handler import get_current_user
from app.database.database import get_db
from app.schemas.transaction import (
    TransactionCreate,
    TransactionResponse,
)
from app.services.transaction_service import (
    create_transaction,
    get_transactions,
    update_transaction,
    delete_transaction,
)

router = APIRouter(
    prefix="/api/transactions",
    tags=["Transactions"],
)


@router.post("", response_model=TransactionResponse)
def add_transaction(
    transaction: TransactionCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return create_transaction(
        db,
        transaction,
        current_user.id,
    )


@router.get("", response_model=List[TransactionResponse])
def list_transactions(
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return get_transactions(
        db,
        current_user.id,
    )


@router.put("/{transaction_id}", response_model=TransactionResponse)
def edit_transaction(
    transaction_id: int,
    transaction: TransactionCreate,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return update_transaction(
        db,
        transaction_id,
        transaction,
        current_user.id,
    )


@router.delete("/{transaction_id}")
def remove_transaction(
    transaction_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(get_current_user),
):
    deleted = delete_transaction(
        db,
        transaction_id,
        current_user.id,
    )

    if not deleted:
        raise HTTPException(
            status_code=404,
            detail="Transaction not found",
        )

    return deleted
